### PR TITLE
[JENKINS-60449] Fix and edge case of OptionalExtension blowing up Jenkins

### PR DESCRIPTION
**Pull request by Vlatombe**
See [JENKINS-60449](https://issues.jenkins-ci.org/browse/JENKINS-60449).

Includes binaries from https://github.com/ikedam/dependencytest/pull/2

If you have an extension like this
```
@OptionalExtension(requirePlugins = {"dependee"})
public class OptionalDependerExtension {
    private static void foo(Dependee d) {

    }
}
```
and `dependee` is not loaded initially, but you load it dynamically later, this blows up, because when Guice does a method scan and can't resolve `Dependee` because it is not yet in classloader at the time of scanning.

This PR can be reviewed commit by commit.
* Introduce a failing test
* Introduce functional changes fixing the test

### Proposed changelog entries

* Entry 1: Fix an edge case of loading optional dependencies that cause Jenkins to blow up
* ...

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### CODE REVIEW

File: core/src/main/java/hudson/PluginManager.java [hunk 1]
```diff
@@ -948,52 +948,50 @@ public void dynamicLoad(File arc, boolean removeExisting, @CheckForNull List<Plu
 
     @Restricted(NoExternalUse.class)
     public void start(List<PluginWrapper> plugins) throws Exception {
-      try (ACLContext context = ACL.as(ACL.SYSTEM)) {
```
Comment (jglick) [inline core/src/main/java/hudson/PluginManager.java]: For other reviewers: **Hide whitespace changes**

Comment (Vlatombe) [reply core/src/main/java/hudson/PluginManager.java]: Yes, there were too many whitespaces inconsistencies in this method, I ended up reformatting it.

Comment (Vlatombe) [reply core/src/main/java/hudson/PluginManager.java]: https://github.com/jenkinsci/jenkins/pull/4393/files?w=1

**Comment by Vlatombe** [comment]
@jglick Updated the PR with a simpler diff (no more refactoring, and a new failing test instead of updating an existing one)

**Review by halkeye — COMMENTED** [review]
This is so over my head, so as a core pr reviewer I'm going to abstain from this one but its cool to try to figure out what's going on

+1 for tests

File: core/src/main/java/hudson/PluginManager.java [hunk 2]
```diff
@@ -949,23 +949,6 @@ public void dynamicLoad(File arc, boolean removeExisting, @CheckForNull List<Plu
     @Restricted(NoExternalUse.class)
     public void start(List<PluginWrapper> plugins) throws Exception {
       try (ACLContext context = ACL.as(ACL.SYSTEM)) {
-        Jenkins.get().refreshExtensions();
```
Comment (fcojfernandez) [inline core/src/main/java/hudson/PluginManager.java]: isn't the refresh needed anymore? do you know why?

**Review by fcojfernandez — APPROVED** [review]
IIUC, code looks good to me

**Review by oleg-nenashev — APPROVED** [review]
Looks fine though it would be great to run PCT against the patched version

**Comment by jglick** [comment]
> Guice does a method scan and can't resolve `Dependee`

By the way: https://github.com/google/guice/pull/1283

**Review by jglick — APPROVED** [review]
